### PR TITLE
#164796889 linkedIn Authentication

### PR DESCRIPTION
**Pull request by Oluwaseyi000**
#### Description
A user will be able to log in using LinkedIn credentials

#### Type of change
New feature (a non-breaking change which adds functionality)

#### How Has This Been Tested?
N/A

#### Checklist:
my code follows style guidelines for this project

#### PT
[#164796889](https://www.pivotaltracker.com/story/show/164796889)

### CODE REVIEW

**Comment by toluola** [comment]
Please fix the Hound issues. Thanks.

File: src/routes/index.js [hunk 1]
```diff
@@ -1,11 +1,37 @@
...
 
-router.get('/', (req, res) => {
+// router.get('/auth/linkedIn', userController.linkedInAuth);
+
+//and then authenticate as:
+
+router.get(
+  '/auth/linkedin',
+  passport.authenticate('linkedin', { state: 'SOME STATE' }),
+  function(req, res) {
```
Comment (Simplemart17) [inline src/routes/index.js]: You might want to change this to arrow function for consistency.

File: src/config/passport-config.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+passport.use(
+  new LinkedInStrategy(
+    {
+      clientID: LINKEDIN_CLIENTID,
+      clientSecret: LINKEDIN_SECRET,
+      callbackURL: '/welcome',
+      state: true,
+    },
+    (accessToken, refreshToken, profile, done) => {
+      console.log(profile);
```
Comment (Temmyogunbo) [inline src/config/passport-config.js]: No console statements!

File: src/config/passport-config.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+      process.nextTick(() => done(null, profile));
```
Comment (Temmyogunbo) [inline src/config/passport-config.js]: Could you explain why you're using `nextTick` method?

File: src/index.js [hunk 4]
```diff
@@ -18,7 +21,7 @@ app.use(bodyParser.urlencoded({ extended: false }));
 app.use(bodyParser.json());
 app.use(cors());
 
-app.get('/', testRoute);
+app.use('/', route);
```
Comment (Temmyogunbo) [inline src/index.js]: We would be mounting our api on `/api/v1`

File: src/routes/index.js [hunk 5]
```diff
@@ -1,11 +1,37 @@
...
+    // The request will be redirected to LinkedIn for authentication, so this
+    // function will not be called.
+  }
+);
+//the login callback:
+
+router.get(
+  '/welcomse',
```
Comment (Temmyogunbo) [inline src/routes/index.js]: fix the typo.

File: src/routes/index.js [hunk 6]
```diff
@@ -1,11 +1,37 @@
...
+  passport.authenticate('linkedin', function(req, res) {
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/routes/index.js]: fixed

File: src/routes/index.js (hunk 1 above)
Comment (Oluwaseyi000) [reply src/routes/index.js]: fixed

File: src/routes/index.js [hunk 7]
```diff
@@ -1,11 +1,37 @@
...
+
+router.get(
+  '/auth/linkedin',
+  passport.authenticate('linkedin', { state: 'SOME STATE' }),
+  function(req, res) {
+    // The request will be redirected to LinkedIn for authentication, so this
+    // function will not be called.
+  }
+);
+//the login callback:
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/routes/index.js]: fixed

File: src/routes/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/routes/index.js]: fixed

File: src/routes/index.js [hunk 8]
```diff
@@ -1,11 +1,37 @@
 import express from 'express';
+import passport from 'passport';
 
 const router = express.Router();
 
-router.get('/', (req, res) => {
+// router.get('/auth/linkedIn', userController.linkedInAuth);
+
+//and then authenticate as:
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/routes/index.js]: fixed

File: src/index.js [hunk 9]
```diff
@@ -40,6 +43,22 @@ app.use((error, res, next) => {
...
+      callbackURL: `http://localhost:${PORT}/welcome`,
+      scope: ['r_emailaddress', 'r_basicprofile'],
+    },
+    function(accessToken, refreshToken, profile, done) {
+      process.nextTick(function() {
+        return done(null, profile);
+      });
+    }
+  )
+);
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/index.js]: fixed

File: src/index.js [hunk 10]
```diff
@@ -40,6 +43,22 @@ app.use((error, res, next) => {
...
 
+passport.use(
+  new LinkedInStrategy(
+    {
+      clientID: LINKEDIN_CLIENTID,
+      clientSecret: LINKEDIN_SECRET,
+      callbackURL: `http://localhost:${PORT}/welcome`,
+      scope: ['r_emailaddress', 'r_basicprofile'],
+    },
+    function(accessToken, refreshToken, profile, done) {
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/index.js]: fixed

File: src/index.js [hunk 11]
```diff
@@ -40,6 +43,22 @@ app.use((error, res, next) => {
   next();
 });
 
+passport.use(
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/index.js]: fixed

File: src/index.js [hunk 12]
```diff
@@ -3,10 +3,13 @@ import bodyParser from 'body-parser';
 import cors from 'cors';
 import dotenv from 'dotenv';
 import swaggerUI from 'swagger-ui-express';
-import testRoute from './routes/index';
+import route from './routes/index';
 import docs from '../swagger.json';
+import passport from 'passport';
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/index.js]: fixed

File: src/config/passport-config.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    }
+  )
+);
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/config/passport-config.js]: fixed

File: src/config/passport-config.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
+import passport from 'passport';
+import LinkedInStrategy from 'passport-linkedin-oauth2';
+import dotenv from 'dotenv';
+
+dotenv.config();
+const { LINKEDIN_CLIENTID, LINKEDIN_SECRET } = process.env;
+
+passport.use(
```
[the remark this answers is not in the document]
Comment (Oluwaseyi000) [reply src/config/passport-config.js]: fixed

**Comment by Oluwaseyi000** [comment]
> Please fix the Hound issues. Thanks.

thanks for the feedback, it's been attended to

**Pull request closed without merging**